README.md
# Draft Post Disclaimer

## About
This plugin is based on work by [Sal Ferrarello](https://salferrarello.com/draft-blog-posts/). The idea being to publish draft posts in a new category and add a disclaimer. This reduces the barrier to getting posts published but lets people know that the post is a work in progress.

### TODO:

 - Remove from Feed? (Make this an option)
 - Turn category select into a drop down
 - Add span/div to front page to display a "Draft" flag if front page display is enabled
 - Add a bit more to documentation
 - Add draft category on init?
 - Fix linting issues
 - Fix issue where PSR-4 auto load only works with composer dump-autoload --optimize 


CHANGELOG.md
# Changelog
All notable changes to this project will be documented in this file.

## [1.0.7] - 2020-02-28
### Fixed
- Eliminated static method calls src/Disclaimer.php #3
- Minor style/linting clean up

## [1.0.6] - 2020-02-28
### Added
- Added composer autoloader
- Added vendor to .gitignore
- Updated namespace and instantion of Options class to allow autoloading

## [1.0.5] - 2020-02-28
### Added
- Added settings link to plugin page. Thanks Sal!

### Fixed
- Moved the settings admin link to the Settings submenu page. Also thanks Sal!

## [1.0.4] - 2020-02-16
### Fixed
- Fixed bug in calling dpd_disclaimer_on_queries_custom_data instead of dpd_exclude_from_queries_custom_data
- Fixed minor spacing to make things more consistent.

### Changed
- Changed static method calls $this style

## [1.0.3] - 2020-02-16
### Added
- Option for excluding draft posts from the main post query (latest posts, archives, etc.)
- Option to display the disclaimer in the main post query (latest posts, archives, etc.)
- Option to toggle disclaimer display on single pages

## [1.0.2] - 2020-02-16
### Fixed
- Fixed reference to old class name causing fatal error.

## [1.0.1] - 2020-02-16
### Updated
- Updated filenames to match class names
- Updated some class names to clean things up a bit
- Updated plugin url to the git repo

### Added
- Added minimum php & wp version
- Added a changelog

### Removed
- Removed todo notes in favor of keeping them in the README
- Removed unnecessary documentation

### Fixed
- Fixed spacing/formatting
- Fixed some typos

## [1.0.0] - 2020-02-15
### Added
- Initial commit.


src/Disclaimer.php
<?php

/**
 * The file that defines the core plugin class
 *
 * @link       http://chrispian.com
 * @since      1.0.0
 *
 * @package    Draft_Post_Dislcaimer
 * @subpackage Draft_Post_Dislcaimer/includes
 */

Namespace CHB\DraftPostDisclaimer;

class Disclaimer {

	/**
	 * @author Chrispian H. Burks <chrispian@gmail.com>
	 * @since  2020-02-15
	 *
	 */
	public function run() {
		// Do the things.
		add_filter( 'pre_get_posts', [ $this, 'exclude_draft_category' ] );
		add_filter( 'the_content', [ $this, 'draft_notice' ] );
		add_filter( 'the_excerpt', [ $this, 'draft_notice_excerpt' ] );
		add_action( 'wp_enqueue_scripts', [ $this, 'add_styles' ] );
	}

	/**
	 * @author Chrispian H. Burks <chrispian@gmail.com>
	 * @since  2020-02-15
	 *
	 * @param $query
	 *
	 * @return mixed
	 */
	public function exclude_draft_category( $query ) {
		$category = strtolower(esc_html(get_option('dpd_category')));
		$exclude_from_queries = get_option('dpd_exclude_from_queries');

		if ( ! $exclude_from_queries['exclude_from_queries_html'] ) {
			return $query;
		}

		if ( $query->is_home || $query->is_archive ) {
			$query->set( 'tax_query', [
				[
					'taxonomy' => 'category',
					'field'    => 'slug',
					'terms'    => $category,
					'operator' => 'NOT IN',
				],
			] );
		}
	return $query;
	}

	/**
	 * @author Chrispian H. Burks <chrispian@gmail.com>
	 * @since  2020-02-15
	 *
	 * @param $content
	 *
	 * @return string
	 */
	public function draft_notice( $content ) {

		if ( ! Disclaimer::maybe_display_notice() ) {
			return $content;
		}

		$disclaimer = wp_kses_post(get_option('dpd_disclaimer'));

		$html = "
		<div class=\"alert alert-warning\" role=\"alert\">
			$disclaimer
		</div>
		";

		return "$content " . $html;

	}

	/**
	 * @author Chrispian H. Burks <chrispian@gmail.com>
	 * @since  2020-02-16
	 *
	 * @param $excerpt
	 *
	 * @return string
	 */
	public function draft_notice_excerpt( ) {

		$excerpt = get_the_excerpt();

		if ( ! Disclaimer::maybe_display_notice() ) {
			return $excerpt;
		}

		$disclaimer = wp_kses_post(get_option('dpd_disclaimer'));

		$html = "
		<div class=\"alert alert-warning\" role=\"alert\">
			$disclaimer
		</div>
		";

		return $excerpt . $html;

	}

	/**
	 * @author Chrispian H. Burks <chrispian@gmail.com>
	 * @since  2020-02-15
	 */
	public function add_styles() {
		$plugin_url = plugin_dir_url( __FILE__ );
		wp_enqueue_style( 'draft-post-disclaimer', $plugin_url . '../assets/style.css' );
	}


	/**
	 * @author Chrispian H. Burks <chrispian.burks@webvdevstudios.com>
	 * @since  2020-02-16
	 * @return mixed
	 */
	public function maybe_display_notice() {

		$status                = true;
		$disclaimer            = get_option( 'dpd_disclaimer' );
		$disclaimer_on_queries = get_option( 'dpd_disclaimer_on_queries' );
		$disclaimer_on_single  = get_option( 'dpd_disclaimer_on_single' );

		global $post;

		if ( ! in_category('draft', $post) ) {
			$status = false;
		}

		if ( ! $disclaimer ) {
			$status = false;
		}

		if ( ! $disclaimer_on_queries['disclaimer_on_queries_html'] && ! is_single() ) {
			$status = false;
		}

		if ( ! $disclaimer_on_single['disclaimer_on_single_html'] ) {
			$status = false;
		}


		return $status;

	}

}


src/admin/options.php
<?php
/**
 * Short Description Goes Here
 *
 * @author Chrispian H. Burks <chrispian@gmail.com>
 * @since  2020-02-15
 */

Namespace CHB\DraftPostDisclaimer;

class Options {

	/**
	 * @author Chrispian H. Burks <chrispian@gmail.com>
	 * @since  2020-02-15
	 *
	 */
	public function run() {

		// Do the things.
		add_action( 'admin_init', ['CHB\DraftPostDisclaimer\Options', 'settings_init'] );
		add_action( 'admin_menu', ['CHB\DraftPostDisclaimer\Options', 'options_page'] );

	}

	/**
	 * @author Chrispian H. Burks <chrispian@gmail.com>
	 * @since  2020-02-15
	 *
	 * Setup our custom options / settings
	 */
	public function settings_init() {

		// Register our settings.
		register_setting('dpd', 'dpd_category');
		register_setting('dpd', 'dpd_disclaimer');

		// Add out section.
		add_settings_section('header', __('Settings', 'dpd'), ['CHB\DraftPostDisclaimer\Options', 'header'], 'dpd');

		// Add the settings filed.
		add_settings_field('category_html', __('Category', 'dpd'), ['CHB\DraftPostDisclaimer\Options', 'category_html'], 'dpd', 'header', [ 'label_for' => 'category_html', 'class' => 'dpd_row', 'dpd_custom_data' => 'custom', ]);
		add_settings_field('disclaimer_html', __('Disclaimer', 'dpd'), ['CHB\DraftPostDisclaimer\Options', 'disclaimer_html'], 'dpd', 'header', [ 'label_for' => 'category_html', 'class' => 'dpd_row', 'dpd_custom_data' => 'custom', ]);

	}

	public function header( $args ) {

		?>
		<p id="<?php echo esc_attr($args['id']); ?>"><?php esc_html_e('Set your draft category and create a disclaimer.', 'dpd'); ?></p>
		<?php
	}


	/**
	 * @author Chrispian H. Burks <chrispian@gmail.com>
	 * @since  2020-02-15
	 *
	 * @param $args array
	 */
	public function category_html( $args ) {

		// get the value of the setting we've registered with register_setting()
		$category   = get_option('dpd_category');
		// output the field
		?>
		<input name="dpd_category" type="text" value="<?php esc_html_e($category, 'dpd'); ?>" /><br /><br />

		<?php
	}

	/**
	 * @author Chrispian H. Burks <chrispian@gmail.com>
	 * @since  2020-02-15
	 *
	 * @param $args array
	 */
	public function disclaimer_html( $args ) {

		// get the value of the setting we've registered with register_setting()
		$disclaimer = get_option('dpd_disclaimer');
		// output the field
		?>
		<textarea name="dpd_disclaimer" cols="80" rows="10"><?php esc_html_e($disclaimer, 'dpd'); ?></textarea><br />

		<?php
	}

	/**
	 * @author Chrispian H. Burks <chrispian@gmail.com>
	 * @since  2020-02-15
	 */
	public function options_page() {
		// add top level menu page
		add_menu_page( 'Draft Post Disclaimer', 'Draft Post Disclaimer Options', 'manage_options', 'dpd', ['CHB\DraftPostDisclaimer\Options', 'options_page_html'] );
	}

	/**
	 * @author Chrispian H. Burks <chrispian@gmail.com>
	 * @since  2020-02-15
	 */
	public function options_page_html() {

		// check user capabilities
		if ( ! current_user_can('manage_options') ) {
			return;
		}

		if ( isset($_GET['settings-updated']) ) {
			add_settings_error('dpd_messages', 'dpd_message', __('Settings Saved', 'dpd'), 'updated');
		}

		settings_errors('dpd_messages');

		?>
		<div class="wrap">
			<h1><?php echo esc_html(get_admin_page_title()); ?></h1>
			<form action="options.php" method="post">
				<?php
				settings_fields('dpd');
				do_settings_sections('dpd');
				submit_button('Save Settings');
				?>
			</form>
		</div>
		<?php
	}
}


draft-post-disclaimer.php
<?php
/**
 * Plugin Name:       Draft Post Disclaimer
 * Plugin URI:        https://github.com/chrispian/draft-post-disclaimer
 * Description:       This plugin adds a disclaimer on posts in a "draft" category.
 * Version:           1.0.7
 * Requires at least: 4.7
 * Requires PHP:      5.4
 * Author:            Chrispian H. Burks
 * Author URI:        http://chrispian.com
 * License:           GPL-2.0+
 * License URI:       http://www.gnu.org/licenses/gpl-2.0.txt
 * Text Domain:       draft-post-disclaimer
 * Domain Path:       /languages
 *
 * @link              http://chrispian.com
 * @since             1.0.0
 * @package           Draft_Post_Dislcaimer
 *
 *
 */

Namespace CHB\DraftPostDisclaimer;

/**
 * Load up our dependencies.
 */
if ( is_readable( __DIR__ . '/vendor/autoload.php' ) ) {
	require_once __DIR__ . '/vendor/autoload.php';
}

// If this file is called directly, abort.
if ( ! defined( 'WPINC' ) ) {
	die;
}

/**
 * Current plugin version.
 */
define( 'DRAFT_POST_DISCLAIMER_VERSION', '1.0.7' );

/**
 * Fire things up.
 *
 * @since    1.0.0
 */
function run_draft_post_disclaimer() {

	$options = new Admin\Options( 'plugin_action_links_' . plugin_basename( __FILE__ ) );
	$options->run();

	$plugin = new Disclaimer();
	$plugin->run();

}
run_draft_post_disclaimer();


src/admin/Options.php
<?php
/**
 * Short Description Goes Here
 *
 * @author Chrispian H. Burks <chrispian@gmail.com>
 * @since  2020-02-15
 */

Namespace CHB\DraftPostDisclaimer\Admin;

class Options {

	/**
	 * @author Chrispian H. Burks <chrispian@gmail.com>
	 * @since  2020-02-28
	 *
	 */
	public function __construct( $plugin_action_links_hook ) {
	    $this->plugin_action_links_hook = $plugin_action_links_hook;
	}


	/**
	 * @author Chrispian H. Burks <chrispian@gmail.com>
	 * @since  2020-02-15
	 *
	 */
	public function run() {

		// Do the things.
		add_action( 'admin_init', [ $this, 'settings_init' ] );
		add_action( 'admin_menu', [ $this, 'options_page' ] );
		add_filter( $this->plugin_action_links_hook, [ $this, 'add_plugin_settings_link' ] );

	}

	/**
	 * @author Chrispian H. Burks <chrispian@gmail.com>
	 * @since  2020-02-15
	 *
	 * Setup our custom options / settings
	 */
	public function settings_init() {

		// Register our settings.
		register_setting('dpd', 'dpd_category');
		register_setting('dpd', 'dpd_disclaimer');
		register_setting('dpd', 'dpd_exclude_from_queries');
		register_setting('dpd', 'dpd_disclaimer_on_queries');
		register_setting('dpd', 'dpd_disclaimer_on_single');

		// Add out section.
		add_settings_section('header', __('Settings', 'dpd'), [ $this, 'header' ], 'dpd');

		// Add the settings filed.
		add_settings_field('category_html', __('Category', 'dpd'), [ $this, 'category_html' ], 'dpd', 'header', [ 'label_for' => 'category_html', 'class' => 'dpd_row', 'dpd_custom_data' => 'custom', ]);
		add_settings_field('exclude_from_queries_html', __('Exclude From Queries?', 'dpd'), [ $this, 'exclude_from_queries_html' ], 'dpd', 'header', [ 'label_for' => 'exclude_from_queries_html', 'class' => 'dpd_row', 'dpd_exclude_from_queries_custom_data' => 'custom', ]);
		add_settings_field('disclaimer_on_queries_html', __('Display Disclaimer on home, archives etc?', 'dpd'), [ $this, 'disclaimer_on_queries_html' ], 'dpd', 'header', [ 'label_for' => 'disclaimer_on_queries_html', 'class' => 'dpd_row', 'dpd_disclaimer_on_queries_custom_data' => 'custom', ]);
		add_settings_field('disclaimer_on_single_html', __('Display Disclaimer on single post view?', 'dpd'), [ $this, 'disclaimer_on_single_html' ], 'dpd', 'header', [ 'label_for' => 'disclaimer_on_single_html', 'class' => 'dpd_row', 'dpd_disclaimer_on_single_custom_data' => 'custom', ]);
		add_settings_field('disclaimer_html', __('Disclaimer', 'dpd'), [ $this, 'disclaimer_html' ], 'dpd', 'header', [ 'label_for' => 'category_html', 'class' => 'dpd_row', 'dpd_category_custom_data' => 'custom', ]);

	}

	public function header( $args ) {

		?>
		<p id="<?php echo esc_attr($args['id']); ?>"><?php esc_html_e('Set your draft category and create a disclaimer.', 'dpd'); ?></p>
		<?php
	}


	/**
	 * @author Chrispian H. Burks <chrispian@gmail.com>
	 * @since  2020-02-15
	 *
	 * @param $args array
	 */
	public function category_html( $args ) {

		// get the value of the setting we've registered with register_setting()
		$category   = get_option('dpd_category');

		?>

		<input name="dpd_category" type="text" value="<?php esc_html_e($category, 'dpd'); ?>" /><br /><br />

		<?php
	}

	/**
	 * @author Chrispian H. Burks <chrispian@gmail.com>
	 * @since  2020-02-16
	 *
	 * @param $args array
	 */
	public function exclude_from_queries_html( $args ) {

		// get the value of the setting we've registered with register_setting()
		$exclude_from_queries   = get_option('dpd_exclude_from_queries');

		?>

		 <select id="<?php echo esc_attr( $args['label_for'] ); ?>" data-custom="<?php echo esc_attr( $args['dpd_exclude_from_queries_custom_data'] ); ?>" name="dpd_exclude_from_queries[<?php echo esc_attr( $args['label_for'] ); ?>]"
 >
			<option value="1" <?php echo isset( $exclude_from_queries[ $args['label_for'] ] ) ? ( selected( $exclude_from_queries[ $args['label_for'] ], '1', false ) ) : ( '' ); ?>>
				<?php esc_html_e( 'Yes', 'dpd' ); ?>
			</option>

			<option value="0" <?php echo isset( $exclude_from_queries[ $args['label_for'] ] ) ? ( selected( $exclude_from_queries[ $args['label_for'] ], '0', false ) ) : ( '' ); ?>>
				<?php esc_html_e( 'No', 'dpd' ); ?>
			</option>
		</select><br /><br />

		<?php
	}

	/**
	 * @author Chrispian H. Burks <chrispian@gmail.com>
	 * @since  2020-02-16
	 *
	 * @param $args array
	 */
	public function disclaimer_on_queries_html( $args ) {

		// get the value of the setting we've registered with register_setting()
		$disclaimer_on_queries   = get_option('dpd_disclaimer_on_queries');

		?>

		<select id="<?php echo esc_attr( $args['label_for'] ); ?>" data-custom="<?php echo esc_attr( $args['dpd_disclaimer_on_queries_custom_data'] ); ?>" name="dpd_disclaimer_on_queries[<?php echo esc_attr( $args['label_for'] ); ?>]"
		>
			<option value="1" <?php echo isset( $disclaimer_on_queries[ $args['label_for'] ] ) ? ( selected( $disclaimer_on_queries[ $args['label_for'] ], '1', false ) ) : ( '' ); ?>>
				<?php esc_html_e( 'Yes', 'dpd' ); ?>
			</option>

			<option value="0" <?php echo isset( $disclaimer_on_queries[ $args['label_for'] ] ) ? ( selected( $disclaimer_on_queries[ $args['label_for'] ], '0', false ) ) : ( '' ); ?>>
				<?php esc_html_e( 'No', 'dpd' ); ?>
			</option>
		</select><br /><br />

		<?php

	}

	/**
	 * @author Chrispian H. Burks <chrispian@gmail.com>
	 * @since  2020-02-16
	 *
	 * @param $args array
	 */
	public function disclaimer_on_single_html( $args ) {

		// get the value of the setting we've registered with register_setting()
		$disclaimer_on_single   = get_option('dpd_disclaimer_on_single');

		?>

		<select id="<?php echo esc_attr( $args['label_for'] ); ?>" data-custom="<?php echo esc_attr( $args['dpd_disclaimer_on_single_custom_data'] ); ?>" name="dpd_disclaimer_on_single[<?php echo esc_attr( $args['label_for'] ); ?>]"
		>
			<option value="1" <?php echo isset( $disclaimer_on_single[ $args['label_for'] ] ) ? ( selected( $disclaimer_on_single[ $args['label_for'] ], '1', false ) ) : ( '' ); ?>>
				<?php esc_html_e( 'Yes', 'dpd' ); ?>
			</option>

			<option value="0" <?php echo isset( $disclaimer_on_single[ $args['label_for'] ] ) ? ( selected( $disclaimer_on_single[ $args['label_for'] ], '0', false ) ) : ( '' ); ?>>
				<?php esc_html_e( 'No', 'dpd' ); ?>
			</option>
		</select><br /><br />

		<?php

	}

	/**
	 * @author Chrispian H. Burks <chrispian@gmail.com>
	 * @since  2020-02-15
	 *
	 * @param $args array
	 */
	public function disclaimer_html( $args ) {

		// get the value of the setting we've registered with register_setting()
		$disclaimer = get_option('dpd_disclaimer');
		// output the field
		?>
		<textarea name="dpd_disclaimer" cols="80" rows="10"><?php esc_html_e($disclaimer, 'dpd'); ?></textarea><br />

		<?php
	}

	/**
	 * @author Chrispian H. Burks <chrispian@gmail.com>
	 * @since  2020-02-15
	 */
	public function options_page() {
		// add top level menu page
		// add_menu_page( 'Draft Post Disclaimer', 'Draft Post Disclaimer Options', 'manage_options', 'dpd', [ $this, 'options_page_html' ] );
		add_submenu_page( 'options-general.php', 'Draft Post Disclaimer Options', 'Draft Post Disclaimer', 'manage_options', 'dpd', [$this, 'options_page_html'] );
	}

	/**
	 * @author Chrispian H. Burks <chrispian@gmail.com>
	 * @since  2020-02-15
	 */
	public function options_page_html() {

		// check user capabilities
		if ( ! current_user_can('manage_options') ) {
			return;
		}

		if ( isset($_GET['settings-updated']) ) {
			add_settings_error('dpd_messages', 'dpd_message', __('Settings Saved', 'dpd'), 'updated');
		}

		settings_errors('dpd_messages');

		?>
		<div class="wrap">
			<h1><?php echo esc_html(get_admin_page_title()); ?></h1>
			<form action="options.php" method="post">
				<?php
				settings_fields('dpd');
				do_settings_sections('dpd');
				submit_button('Save Settings');
				?>
			</form>
		</div>
		<?php
	}

	/**
	 * @author Chrispian H. Burks <chrispian.burks@webvdevstudios.com>
	 * @since  2020-02-28
	 *
	 * @param $links
	 *
	 * @return mixed
	 */
	public function add_plugin_settings_link( $links ) {
	    $settings_link = '<a href="options-general.php?page=dpd">' . __( 'Settings' ) . '</a>';
	    array_push( $links, $settings_link );
	  	return $links;
	}

}
